File: Neetcode/Python/Two-Pointers/3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums = sorted(nums)
        res = []

        for left in range(0,len(nums)-2):
            i=left+1
            right = len(nums)-1
            while i<right and i>left:
                sums = nums[left] + nums[i] + nums[right]
                if sums == 0:
                    res.append([nums[left],nums[i],nums[right]])
                    i += 1
                elif sums > 0:
                    right -= 1
                else:
                    i += 1
        res = set(map(tuple,res))
        return list(res)

File: Neetcode/Golang/Two-Pointers/3sum.go
package main

import "sort"

func threeSum(nums []int) [][]int {
	sort.Ints(nums)
	var res [][]int

	for num1Idx := 0; num1Idx < len(nums)-2; num1Idx++ {
		if num1Idx > 0 && nums[num1Idx] == nums[num1Idx-1] {
			continue
		}
		l, r := num1Idx+1, len(nums)-1
		for l < r {
			threeSum := nums[num1Idx] + nums[l] + nums[r]
			if threeSum == 0 {
				res = append(res, []int{nums[num1Idx], nums[l], nums[r]})
				r--
				for l < r && nums[r] == nums[r+1] {
					r--
				}
			} else if threeSum > 0 {
				r--
			} else {
				l++
			}
		}
	}
	return res
}
